buildpacks/pack
# Enable caching layers in image registry via `cache-image` flag

## Summary
Adds a `--cache-image <cache-image-name>` flag that will push all `cache=true` layers to the `cache-image-name` image. This PR only supports pushing to a remote registry (so the `--publish` flag must be used as well).

the `cache-image` will allow subsequent rebuild to fetch these layers and speed up rebuilds.

Note that only remote cache images are supported at this time. (Local support would need a lifecycle change)

## Documentation



- Should this change be documented?
    - [x] Yes, see https://github.com/buildpacks/docs/issues/266

## Related
This change will through some errors ( though they will not impact program execution ) until [this](https://github.com/buildpacks/lifecycle/commit/c0e33588c4a40640ac228622b6114921e9cd73fc) lifecycle PR is included in a release. 

Resolves #491 
Resolves #166


## Validation Steps
To validate this PR you will need to 
1) build a up to date version of the lifecycle
  - clone the `github.com/buildpacks/lifecycle` repository & run a `git pull` (make sure you are on the `main` branch)
  - build a version of the lifecycle by running `LIFECYCLE_VERSION=0.9.4 make build-image-linux` inside of the lifecycle repository
2) create a custom builder
  - Use the following example `builder.toml` file at the bottom of this comment
  - replace the `uri = "<uri>"` under the `lifecycle` heading with `uri = "<path to the lifecycle-repo/out/lifecycle-v0.9.4+linux.x86-64.tgz>"`
  -  run `pack create-builder builder-test --config builder.toml --pull-policy if-not-present` to create a builder
3) trust the created builder so that we can use the `--publish` flag when testing
  - run `pack add-trusted-builder builder-test`
4) run a registry locally
  - run a local registry with `docker run -e REGISTRY_STORAGE_DELETE_ENABLED=true -d -p 5000:5000 --restart always --name registry registry:2`
5) run a pack build with the local registry
  - recommend using the `https://github.com/paketo-buildpacks/samples/java/maven` sample application for building.
  - if on a OSX system substitute `localhost` for `docker.for.mac.localhost` in the below command.
  - `./out/pack build localhost:5000/test/maven-example -p <path/to/app> --publish --cache-image localhost:5000/test/maven-example.cache --builder  tag-builder-test`
6) Rebuild and observer that the `restorer` pulls cache layers from the previous cache image
  - run  `./out/pack build localhost:5000/test/maven-example2 -p <path/to/app> --publish --cache-image localhost:5000/test/maven-example.cache --builder  tag-builder-test`


## Validation notes
If during steps `5` and `6` above we build and push to the `dockerhub` repository, our upload will time out with the following error: 
```
Adding cache layer 'paketo-buildpacks/maven:cache'
Warning: Failed to export cache: committing cache: saving image 'index.docker.io/donthedinosaur/some-java-image.build:latest': failed to write image to the following tags: [index.docker.io/donthedinosaur/some-java-image.build:latest: PATCH https://index.docker.io/v2/donthedinosaur/some-java-image.build/blobs/uploads/21145335-2d37-4b98-8c76-2e2ff3a0f6cc?_state=REDACTED: unexpected status code 504 Gateway Timeout: <html><body><h1>504 Gateway Time-out</h1>
The server didn't respond in time.
</body></html>

]
```



##### Builder.toml file to copy
```
description = "Test image for pack PR970"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/dotnet-core:0.0.9"
  version = "0.0.9"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/go:0.2.8"
  version = "0.2.8"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/java-native-image:4.5.0"
  version = "4.5.0"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/java:4.5.0"
  version = "4.5.0"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/nginx:0.0.194"
  version = "0.0.194"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/nodejs:0.0.10"
  version = "0.0.10"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/procfile:3.0.0"
  version = "3.0.0"

[[buildpacks]]
  image = "gcr.io/paketo-buildpacks/ruby:0.1.7"
  version = "0.1.7"

[lifecycle]
  version = "0.10.1"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/ruby"
    version = "0.1.7"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core"
    version = "0.0.9"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/nodejs"
    version = "0.0.10"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/go"
    version = "0.2.8"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/nginx"
    version = "0.0.194"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/java-native-image"
    version = "4.5.0"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/java"
    version = "4.5.0"

[[order]]

  [[order.group]]
    id = "paketo-buildpacks/procfile"
    version = "3.0.0"

[stack]
  id = "io.buildpacks.stacks.bionic"
  build-image = "docker.io/paketobuildpacks/build:1.0.7-base-cnb"
  run-image = "index.docker.io/paketobuildpacks/run:base-cnb"
  run-image-mirrors = ["gcr.io/paketo-buildpacks/run:base-cnb"]



```
